### DISCUSSION TITLE
False Positive - CVE-2023-1370 - Getting detected even after upgrading json-smart to 2.4.10

### DISCUSSION THREAD

**Discussion by avinashkr2712** [aquasecurity/trivy#4114] [Q&A]
## Description

  Library                                        │ Vulnerability      │ Severity │ Installed Version │ Fixed Version

 net.minidev:json-smart (app.jar) │ CVE-2023-1370 │ HIGH     │ 2.4.8                     │ 2.4.9

Trivy is giving false positive even after updating to Fixed version.

**Comment by knqyf263** [comment]
It is possible to have multiple same packages in a single JAR file. Please make sure you bumped all of them.

**Comment by avinashkr2712** [comment]
@knqyf263 Thanks for your response!! I have already done above step , I have also excluded the transitive dependency and added a explicit version of 2.4.10 , Still I am getting this vulnerability in my image. Am I missing something here ?

**Comment by DmitriyLewen** [comment]
Hello @avinashkr2712 
Can you scan your image with `-f json --list-all-pkgs` flag?
Check path to `json-smart 2.4.8`, perhaps another jar contains this package.

> I have also excluded the transitive dependency and added a explicit version of 2.4.10

Can you share your `jar` file(you can exclude all `*.class` files from file)?
Maybe one of nested `jars` contains this package. 

Regards, Dmitriy

**Comment by ctrung** [comment]
FYI, it happens with [shaded jars](https://maven.apache.org/plugins/maven-shade-plugin/examples/class-relocation.html). 

Eg. with nimbus-jose-jwt or msal described here that i also encountered recently : 
https://github.com/Azure/azure-sdk-for-java/issues/34223#issuecomment-1487227114

**Reply by DmitriyLewen** [reply]
Hello @ctrung 
Thanks for your report!

I investigated `nimbus-jose-jwt-9.23.jar` and understand this case.

There is 1 problem - we can't get version of shaded dependency.
`nimbus-jose-jwt-9.23/META-INF/maven/net.minidev/json-smart/pom.properties` has old version(`2.4.8`),
`nimbus-jose-jwt-9.23/META-INF/maven/com.nimbusds/nimbus-jose-jwt/pom.xml` just shows that source code of `json-smart` has been changed.
e.g. in `nimbus-jose-jwt-9.23.jar` source code of `json-smart` has been changed changed, but we can't find version of those changes(`2.4.10. in this case).

Also we can't skip those dependencies, because shaded dependency might change another part of the source code (I mean these changes don't fix the vulnerability).

That is why we should show shaded dependencies.

To fix special cases we have filtering options. E.g. [OPA](https://aquasecurity.github.io/trivy/v0.40/docs/vulnerability/examples/filter/#by-open-policy-agent) or [Modules](https://aquasecurity.github.io/trivy/v0.40/docs/advanced/modules/).

Best Regards, Dmitriy

**Comment by mqvmai** [comment]
@avinashkr2712 Was there ever any resolution to this issue?  I tried upgrading msal4j library, which does fix the vulnerability but that comes with some other build issues.
